packages/discovery-react-components/src/utils/document/documentUtils.ts

```diff
@@ -28,27 +28,38 @@ export function findOffsetInDOM(
 ): FindOffsetResults {
   const allNodes = Array.from(parentNode.querySelectorAll('[data-child-begin]')) as HTMLElement[];
 
-  let beginNode,
-    endNode,
-    idx = 0;
-
-  do {
-    beginNode = allNodes[idx];
-  } while (
-    ++idx &&
-    idx < allNodes.length &&
-    begin >= parseInt(allNodes[idx].dataset.childBegin || '0', 10)
-  );
+  let beginNode = null,
+    endNode = null,
+    beginNodeSize = Number.MAX_VALUE,
+    endNodeSize = Number.MAX_VALUE;
+  allNodes.forEach((node: HTMLElement) => {
```

Comment:
This will run through all of the nodes every time, even if the matches were at the beginning of the array. This should break out of the loop once we've found the correct matches.

Reply:
We won't know that we've hit the smallest matches (i.e. the nodes lowest down the DOM tree) until we've checked every node, so we do need to check all the nodes. We could chop off the first chunk of nodes that have too small of `begin` values (using a binary search since the nodes are inherently sorted by `begin`), but I wasn't sure if the performance gains would be enough to be worth the code complexity. (The procedure would still be O(n), but this optimization would likely make it more efficient for larger documents and / or small highlights.)